L1Trigger/L1TTrackMatch/plugins/L1TkHTMissEmulatorProducer.cc

```diff
@@ -0,0 +1,286 @@
+
+/**\class L1TrackerHTMissEmulatorProducer L1TrackerHTMissEmulatorProducer.cc
+ L1Trigger/L1TTrackMatch/plugins/L1TrackerHTMissEmulatorProducer.cc
+ Description: Takes L1TTkJets and performs a integer emulation of Track-based missing HT, outputting a collection of EtSum 
+*/
+
+// Original Author:  Hardik Routray
+//         Created:  Mon, 11 Oct 2021
+
+// system include files
+#include <memory>
+#include <numeric>
+
+// user include files
+#include "FWCore/Framework/interface/Frameworkfwd.h"
+#include "FWCore/Framework/interface/stream/EDProducer.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+#include "FWCore/Framework/interface/ESHandle.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "DataFormats/Math/interface/LorentzVector.h"
+#include "DataFormats/L1TCorrelator/interface/TkPrimaryVertex.h"
+#include "DataFormats/L1TCorrelator/interface/TkHTMiss.h"
+#include "DataFormats/L1TCorrelator/interface/TkHTMissFwd.h"
+#include "DataFormats/L1Trigger/interface/EtSum.h"
+#include "DataFormats/L1Trigger/interface/TkJetWord.h"
+#include "L1Trigger/L1TTrackMatch/interface/L1TkHTMissEmulatorProducer.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+
+using namespace l1t;
+
+class L1TkHTMissEmulatorProducer : public edm::stream::EDProducer<> {
```

Comment:
not quite sure how to go about implementing this. Currently, all the LUTs are output to a separate directory named LUTs if `saveLUTs = cms.bool(True)`

Reply:
But, if there are multiple threads calling the writeLUTToFile function (https://github.com/cms-l1t-offline/cmssw/pull/957/files#diff-fc15aeaa0480bfb1c4a7e9a4d540c578c7ac83e29dcb184b99d7bed445ac6a6cR75-R77) then all but one of them will return an exception. This shouldn't be happening. I can think of two ways to handle this, one obviously more complex than the other:
1. **C++**: Create a lock that only allows one stream to write to a file/directory at a time (i.e. only one can access `writeLUTtoFile` as any given time). You will also need either different directories or different files for each stream since allowing them to write to the same file would not be deterministic, even with the lock. Plus, I think they would all write the same information to the files, so your just duplicating information by `nstreams`.
2. **Python**: Sometime after [this line in L1TrackObjectNtupleMaker_cfg.py](https://github.com/cms-l1t-offline/cmssw/blob/84f7870b0ee693b34af40fdb85f9bab0add48ba7/L1Trigger/L1TTrackMatch/test/L1TrackObjectNtupleMaker_cfg.py#L81) you could add some code that will simply turn off the saving of LUTs if there are multiple streams. It doesn't really protect the C++, which is a big downside, but it does prevent this from happening in this one instance.
```python
if nstreams > 1:
    process.L1TrackerEmuHTMiss.saveLUTs = cms.bool(False)
```
Perhaps one of the CMSSW developers will know a better way.